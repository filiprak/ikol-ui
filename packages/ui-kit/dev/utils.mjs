import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
