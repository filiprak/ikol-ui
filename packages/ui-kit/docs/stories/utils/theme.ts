import { useTheme } from '@ui/composables';
import { createIkolUI } from '@ui/index';
import { App } from 'vue';
import { StoryContext, VueRenderer } from '@storybook/vue3';

export function setupTheme(context: any) {
    const theme = useTheme();

    theme.is_dark.value = context?.globals?.theme === 'dark';

    return theme;
}

export function setupApp(app: App, context?: StoryContext<VueRenderer>) {
    app.use(createIkolUI());
}
