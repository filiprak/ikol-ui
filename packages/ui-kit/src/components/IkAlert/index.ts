import IkAlert from './IkAlert.vue';
export interface IkAlert extends InstanceType<typeof IkAlert> { };
export { IkAlert };
