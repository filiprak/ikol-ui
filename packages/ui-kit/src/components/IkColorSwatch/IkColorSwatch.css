.ik-color-swatch {
    width: 32px;
    height: 32px;
    border: 1px solid #ccc;
}

.ik-color-swatch--round {
    border-radius: 50%;
}

.ik-color-swatch--xs {
    width: 12px;
    height: 12px;
}

.ik-color-swatch--sm {
    width: 16px;
    height: 16px;
}

.ik-color-swatch--md {
    width: 20px;
    height: 20px;
}

.ik-color-swatch--lg {
    width: 24px;
    height: 24px;
}