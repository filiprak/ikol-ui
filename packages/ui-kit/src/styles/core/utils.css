/* shadow */
.ik-shadow-0 {
    box-shadow: none;
}

.ik-shadow-1 {
    box-shadow: 0 1px 3px 0 var(--shadow-color-1);
}

.ik-shadow-2 {
    box-shadow: 0 2px 5px 0 var(--shadow-color-1);
}

.ik-shadow-3 {
    box-shadow: 0 2px 10px 0 var(--shadow-color-1);
}

.ik-shadow-4 {
    box-shadow: 0 2px 15px 0 var(--shadow-color-1);
}

.ik-shadow-5 {
    box-shadow: 0 2px 20px 0 var(--shadow-color-1);
}

.ik-shadow-6 {
    box-shadow: 0 3px 30px 0 var(--shadow-color-1);
}

/* skeleton background */
@-webkit-keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

@keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

.ik-bg-skeleton {
    background: var(--skeleton-color-1);
    overflow: hidden;
    position: relative;
    user-select: none;
}

.ik-bg-skeleton:after {
    background: linear-gradient(90deg, hsla(0, 0%, 100%, 0), var(--skeleton-color-2), hsla(0, 0%, 100%, 0));
    -webkit-animation: skeleton-loading 1.5s infinite;
    animation: skeleton-loading 1.5s infinite;
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
    z-index: 1
}

.ik-cursor--pointer {
    cursor: pointer;
}

/* design utils */
.ik-design--bg-primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-design--bg-success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-design--bg-error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-design--bg-accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-design--bg-secondary {
    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
}

.ik-design--bg-default {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);
}

/* height */
.ik-height--full {
    height: 100%;
}

.ik-width--full {
    width: 100%;
}
