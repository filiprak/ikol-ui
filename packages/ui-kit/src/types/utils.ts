import type { ExtractPropTypes } from 'vue';

/* eslint-disable @typescript-eslint/naming-convention */
export type IkExtractProps<T> =
    T extends new (...angs: any) => { $props: infer P; } ? NonNullable<P> : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: infer P, ...args: any) => any ? P : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkExtractPropTypes<T> = Prettify<Readonly<ExtractPropTypes<T>>>;

export type IkExtractExposed<T> =
    T extends new (...angs: any) => infer E ? E : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: any, ctx: any, expose: (exposed: infer E) => any, ...args: any) => any ? NonNullable<E> : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkKeysByType<T, U> = keyof {
    [K in keyof T as T[K] extends U ? K : never]: T[K];
};

export enum IkUIDesignColor {
    primary = 'primary',
    success = 'success',
    error = 'error',
    accent = 'accent',
    default = 'default',
}

export enum IkUIElemSize {
    xs = 'xs',
    sm = 'sm',
    md = 'md',
    lg = 'lg',
}

export enum IkUIVariant {
    flat = 'flat',
    filled = 'filled',
    outline = 'outline',
    pill = 'pill',
}

export enum IkPosition {
    top = 'top',
    right = 'right',
    bottom = 'bottom',
    left = 'left',
    horizontal = 'horizontal',
    vertical = 'vertical',
}

export enum IkAlignment {
    start = 'start',
    center = 'center',
    end = 'end',
}

export type EnumT = Record<string, string> | Record<string | number, string | number>;
export type ObjValueT<T> = T[keyof T];
export type DefinedKeys<T> = { [K in keyof T]: undefined extends T[K] ? never : K }[keyof T];
export type OmitUndefined<T> = { [K in DefinedKeys<T>]: T[K] };
export type NotArray<T> = T extends any[] ? never : T; // eslint-disable-line @typescript-eslint/no-explicit-any
export type DefinedComponent = new (...args: any[]) => any; // eslint-disable-line @typescript-eslint/no-explicit-any
export type Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
