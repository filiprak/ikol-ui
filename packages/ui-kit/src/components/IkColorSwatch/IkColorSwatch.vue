<template>
    <div :class="root_class"
         :style="root_style"></div>
</template>

<script setup lang="ts">

import './IkColorSwatch.css';
import type { IkUIElemSize } from '@ui/types/utils';

import { computed } from 'vue';

const props = withDefaults(defineProps<{
    format?: 'css' | 'hex' | 'rgb' | 'rgba',
    color?: string | string[]
    round?: boolean,
    size?: `${IkUIElemSize}`
}>(), {
    format: 'hex',
    round: false,
});

const root_class = computed(() => {
    const classes: Record<string, boolean> = {
        'ik-color-swatch': true,
        'ik-color-swatch--round': props.round,
    };
    if (props.size) {
        classes['ik-color-swatch--' + props.size] = !!props.size;
    }
    return classes;
});

const root_style = computed(() => {
    if (props.color) {
        let color_style = null;
        switch (props.format) {
            case 'css':
                color_style = props.color as string;
                break;
            case 'hex':
                color_style = '#' + props.color;
                break;
            case 'rgb':
                color_style = (typeof props.color === 'string') ? props.color : 'rgb(' + props.color.join(', ') + ')';
                break;
            case 'rgba':
                color_style = (typeof props.color === 'string') ? props.color : 'rgba(' + props.color.join(', ') + ')';
                break;
        }
        if (!color_style) {
            return {};
        }
        return { backgroundColor: color_style };
    } else {
        return {};
    }
});

</script>
