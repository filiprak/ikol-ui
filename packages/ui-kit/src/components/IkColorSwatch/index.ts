import IkColorSwatch from './IkColorSwatch.vue';
export interface IkColorSwatch extends InstanceType<typeof IkColorSwatch> { };
export { IkColorSwatch };
