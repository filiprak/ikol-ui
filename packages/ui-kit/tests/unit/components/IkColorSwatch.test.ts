import { renderDeep } from '@ui/tests/tools/utils';
import { IkColorSwatch } from '@ui/components/IkColorSwatch';

describe('IkColorSwatch', () => {
    const getInstance = (options = {}) => renderDeep(IkColorSwatch, options);

    it('has correct classes, style and default props', () => {
        const wrapper = getInstance();
        expect(wrapper.vm.format).toEqual('hex');
        expect(wrapper.vm.color).toEqual(undefined);
        expect(wrapper.vm.round).toEqual(false);

        expect(wrapper.vm.root_class).toEqual({
            'ik-color-swatch': true,
            'ik-color-swatch--round': false,
        });
        expect(wrapper.vm.root_style).toEqual({});
        expect(wrapper.element).toMatchSnapshot();
    });

    it('has correct classes and style when round', () => {
        const wrapper = getInstance({
            propsData: {
                round: true,
            },
        });
        expect(wrapper.vm.root_class).toEqual({
            'ik-color-swatch': true,
            'ik-color-swatch--round': true,
        });
        expect(wrapper.vm.root_style).toEqual({});
        expect(wrapper.element).toMatchSnapshot();
    });

    it('has correct style for hex color', () => {
        const wrapper = getInstance({
            propsData: {
                round: true,
                format: 'hex',
                color: 'ff00aa',
            },
        });
        expect(wrapper.vm.root_style).toEqual({ backgroundColor: '#ff00aa' });
        expect(wrapper.element).toMatchSnapshot();
    });

    it('has correct style for rgb format', () => {
        const wrapper = getInstance({
            propsData: {
                round: true,
                format: 'rgb',
                color: 'rgb(0.1, 0.0, 1.0)',
            },
        });
        expect(wrapper.vm.root_style).toEqual({ backgroundColor: 'rgb(0.1, 0.0, 1.0)' });
    });

    it('has correct style for rgb array format', () => {
        const wrapper = getInstance({
            propsData: {
                round: true,
                format: 'rgb',
                color: [0.1, 0.0, 1.0],
            },
        });
        expect(wrapper.vm.root_style).toEqual({ backgroundColor: 'rgb(0.1, 0, 1)' });
    });

    it('has correct style for rgba format', () => {
        const wrapper = getInstance({
            propsData: {
                round: true,
                format: 'rgba',
                color: 'rgba(0.1, 0.0, 1.0, 1.0)',
            },
        });
        expect(wrapper.vm.root_style).toEqual({ backgroundColor: 'rgba(0.1, 0.0, 1.0, 1.0)' });
    });

    it('has correct style for rgba array format', () => {
        const wrapper = getInstance({
            propsData: {
                round: true,
                format: 'rgba',
                color: [0.1, 0.0, 1.0, 1.0],
            },
        });
        expect(wrapper.vm.root_style).toEqual({ backgroundColor: 'rgba(0.1, 0, 1, 1)' });
    });

    it('has correct style for css format', () => {
        const wrapper = getInstance({
            propsData: {
                round: true,
                format: 'css',
                color: '#3344ff',
            },
        });
        expect(wrapper.vm.root_style).toEqual({ backgroundColor: '#3344ff' });
    });

    it('renders with custom size', () => {
        const wrapper = getInstance({
            propsData: {
                size: 'lg',
            },
        });
        expect(wrapper.classes()).toContain('ik-color-swatch--lg');
    });
});
