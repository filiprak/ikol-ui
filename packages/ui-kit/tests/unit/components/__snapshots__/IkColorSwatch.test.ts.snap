// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkColorSwatch has correct classes and style when round 1`] = `
<div
  class="ik-color-swatch ik-color-swatch--round"
/>
`;

exports[`IkColorSwatch has correct classes, style and default props 1`] = `
<div
  class="ik-color-swatch"
/>
`;

exports[`IkColorSwatch has correct style for hex color 1`] = `
<div
  class="ik-color-swatch ik-color-swatch--round"
  style="background-color: rgb(255, 0, 170);"
/>
`;
